xbmc/addons/BinaryAddonCache.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ *      Copyright (C) 2016 Team Kodi
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "BinaryAddonCache.h"
+#include "AddonManager.h"
+#include "threads/SingleLock.h"
+
+namespace ADDON
+{
+
+CBinaryAddonCache::~CBinaryAddonCache()
+{
+  CAddonMgr::GetInstance().UnregisterObserver(this);
+}
+
+void CBinaryAddonCache::Init()
+{
+  Update();
+  CAddonMgr::GetInstance().RegisterObserver(this);
```

Comment:
Shouldn't this be in reverse order? You might miss an update right between `Update();` finishing and `RegisterObserver()` being called.